### PR TITLE
Add locations for `else`, `ensure`, `end` keywords

### PR DESCRIPTION
**Pull request by FnControlOption**
This adds some `Location` fields to `If`, `Expressions`, and `ExceptionHandler` nodes for keywords `else`, `ensure`, etc. and also fills the empty `location` field of `Rescue` nodes.

These will be helpful in [Ameba](https://github.com/crystal-ameba/ameba) to report style issues. For example:

- https://github.com/crystal-ameba/ameba/pull/254
  ```cr
  # bad
  def test
    if something
      work
    end
  end

  # good
  def test
    return unless something
    work
  end
  ```
  ```cr
  # good
  def test
    if something
      work
    else
      # nothing
    end
  end
  ```
- https://github.com/crystal-ameba/ameba/blob/63a6c73/src/ameba/rule/style/redundant_begin.cr#L106-L154
  ```cr
  # bad
  def method
    begin
      read_content
    rescue
      close_file
    end
  end

  # good
  def method
    read_content
  rescue
    close_file
  end
  ```

### CODE REVIEW

**Review by straight-shoota — APPROVED** [review]

**Comment by beta-ziliani** [comment]
I wonder if `end_keyword_location` shouldn't just be calculated from `end_location`. I would expect the following equation to always hold: `end_keyword_location.line == end_location.line && end_keyword_location.column == end_location.column - 2`.

**Comment by straight-shoota** [comment]
Good point. But I don't think it's true in the general sense. `Expressions` can be delimited by parenthesis, so the end keyword would be `)`. That should probably be added to this PR.

**Comment by beta-ziliani** [comment]
Good point about the (). What do you mean by "that" in "that should be added to this PR"?

**Comment by straight-shoota** [comment]
"that" end keyword location gets assigned to `Expressions` with parenthesis as well.

**Comment by beta-ziliani** [comment]
@FnControlOption can you add the relevant locations to `()` as well?

**Comment by FnControlOption** [comment]
Finally got around to working on this again... Deleted `begin_keyword_location` and `end_keyword_location` because they're basically the same as `location` and `end_location`. (We can use `Expressions#keyword` to check if the block is enclosed by `begin/end` or by parentheses.)

**Review by straight-shoota — APPROVED** [review]

**Review by straight-shoota — CHANGES_REQUESTED** [review]
Actually, no this doesn't seem to be right.

There are already properties for `else` and `ensure`, they're just missing their location. I don't think we need separate location properties, just apply the locations to the respective sub nodes.

**Comment by FnControlOption** [comment]
Thanks, I didn't think of that! The changes are much nicer now.

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -1327,23 +1327,24 @@ module Crystal
...
+      begin_location = @token.location
       slash_is_regex!
       next_token_skip_statement_end
       exps = parse_expressions
-      node, end_location = parse_exception_handler exps
+      node, end_location = parse_exception_handler exps, begin_location: begin_location
       node.end_location = end_location
       if !node.is_a?(ExceptionHandler) && (!node.is_a?(Expressions) || node.keyword)
-        node = Expressions.new([node]).at(node).at_end(node)
+        node = Expressions.new([node]).at(begin_location).at_end(end_location)
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr:1337]: I presume this change shouldn't be necessary. `at` and `at_end` take the start and end locations from `node` which should already be set appropriately.

Comment (FnControlOption) [reply src/compiler/crystal/syntax/parser.cr:1337]: This is needed for:

```crystal
begin
  foo
end
```

because `node` is the AST node for `foo`

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
There are some failing playground specs now. I suppose they just need the line numbers adjusted.

**Comment by FnControlOption** [comment]
Wait, maybe we shouldn't change the `location` of `else` because it might not be an `Expressions` node

**Comment by HertzDevil** [comment]
This looks wrong now. The location of the else-clause is necessarily different from the `else` keyword.

The grammar change should really be handled as part of #5258.

**Comment by asterite** [comment]
I don't think we should add `Expressions` nodes to wrap simple nodes. It's more memory consumed, and a slightly slower compiler.

Instead, we should add properties like `else_location` and `ensure_location`, etc.

**Comment by straight-shoota** [comment]
Oh, yeah that's a good point.
I suppose we should revert back to 720d4d9 then?

File: src/compiler/crystal/syntax/ast.cr [hunk 2]
```diff
@@ -700,6 +703,9 @@ module Crystal
     property then : ASTNode
     property else : ASTNode
 
+    # The location of the 'else' keyword if present.
```
Comment (Sija) [inline src/compiler/crystal/syntax/ast.cr:706]:
```suggestion
    # The location of the `else` keyword if present.
```

File: src/compiler/crystal/syntax/ast.cr [hunk 3]
```diff
@@ -1610,6 +1616,12 @@ module Crystal
     property implicit = false
     property suffix = false
 
+    # The location of the 'else' keyword if present.
```
Comment (Sija) [inline src/compiler/crystal/syntax/ast.cr:1619]:
```suggestion
    # The location of the `else` keyword if present.
```

File: src/compiler/crystal/syntax/ast.cr [hunk 4]
```diff
@@ -1610,6 +1616,12 @@ module Crystal
...
+    property else_location : Location?
+
+    # The location of the 'ensure' keyword if present.
```
Comment (Sija) [inline src/compiler/crystal/syntax/ast.cr:1622]:
```suggestion
    # The location of the `ensure` keyword if present.
```

File: src/compiler/crystal/syntax/ast.cr [hunk 5]
```diff
@@ -677,6 +677,9 @@ module Crystal
     property else : ASTNode
     property? ternary : Bool
 
+    # The location of the 'else' keyword if present.
```
Comment (Sija) [inline src/compiler/crystal/syntax/ast.cr:680]:
```suggestion
    # The location of the `else` keyword if present.
```

**Review by Sija — COMMENTED** [review]
OCD-fueled suggestion

**Comment by straight-shoota** [comment]
For the future: Please don't force push changes to pending PRs. That makes it unnecessarily hard for reviewers. Just append new commits. We're squashing on merge anyways. Ref: https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#making-good-pull-requests

**Review by beta-ziliani — APPROVED** [review]
Thanks @FnControlOption 🙇

**Pull request merged**